activities/writing-unit-tests.md

```diff
@@ -0,0 +1,151 @@
+---
+title: ESWN 2021 - Writing unit tests
+description: Tutorial for 2021 ESWN PyGMT workshop
+slideOptions:
+  theme: solarized
+---
+
+## Writing unit tests
+
+*Because tested code is code that you can trust*
+
+<!-- Put the link to this slide here so people can follow -->
+<small> P.S. Slides are at https://hackmd.io/@pygmt/eswn-2021-writing-unit-tests</small>
+
+---
+
+## Why test?
+
+Imagine choosing two vaccines :syringe::
+
+- Vaccine 1 comes with the exact biochemical equations on how it works
+- How vaccine 2 works is a mystery, but it has been extensively tested on thousands of people, across diverse populations, with little side effects
+
+Which one would you choose :thinking_face:?
+
+---
+
+## Testing code helps to
+
+- Ensure functions work as expected :green_heart: (even if they seem like a black box)
+- Prevent bugs :bug: from popping up over time
+- Give us confidence when refactoring :recycle: (i.e. rewriting) code to be more efficient, while keeping the same functionality
+
+---
+
+## Structure of a unit test 🧪
+
+Let's start with a simple Python function that adds two numbers:
+
+```python
+def add(x, y):
+    return x + y
+```
+
+Now let's check that this function works on two input numbers. We'll use a testing framework/library called [`pytest`](https://docs.pytest.org/en/6.2.x/) that allows us to easily check the output using `assert` statements.
+
+```python
+def test_add():
+    value = add(x=1.0, y=2.0)
+    assert value == 3.0
+```
+
+---
+
+## Where do the tests go?
+
+In a Python project using [`pytest`](https://docs.pytest.org/en/6.2.x/), we usually structure our code in folders like so:
+
+```markdown
+projectname/
+  ├──src/
+  │   └──calc.py (contains 'add' function)
+  └──tests/
+      └──test_calc.py (contains 'test_add' function)
```

Comment:
```suggestion
projectname/
  ├──__init__.py
  ├──src/
  │   └──calc.py (contains 'add' function)
  └──tests/
      ├──test_calc.py (contains 'test_add' function)
      └──__init__.py
```
Not sure if I formatted the tree properly, but `__init__.py` is needed for the example to work.